### ISSUE TITLE
NNUE undo promotion

### ISSUE THREAD

**Issue by syzygy1** [official-stockfish/Stockfish#2952]
https://github.com/official-stockfish/Stockfish/blob/add890a10b8fe03e091520cd0af7383615c6c386/src/position.cpp#L933-L942

Can it be correct that the NNUE board is not updated here? I would think this is a bug...

In general, it would seem possible to move the NNUE-board updating code into the put_piece()/remove_piece()/move_piece() code. (I haven't looked at that in depth yet.)

**Comment by syzygy1** [comment]
Hmmm, probably not a bug. The regular undo code is already putting back the pawn, I suppose....

**Comment by vondele** [comment]
@syzygy1 I assume this can be tested comparing the eval of 
position fen1 moves x y z
eval
position fen2 
eval
where the two positions are the same?

**Comment by syzygy1** [comment]
@vondele After looking a bit better at the code, I realised that the code section I quoted replaces the promoted piece on the **to** square with a pawn. The regular undo code then moves that pawn back to the **from** square. The NNUE code instead just empties the **to** square and places the pawn back to the **from** square, which is fine as well.

So the code seems correct but some things could possibly be done a bit nicer. I don't understand yet what board information the NNUE code requires exactly, though.
